File: 2-strings/java/387-first-unique-character.java
public class Solution{
    public static int firstUniqChar(String s) {
        int[] charCounts = new int[26];

        for (int i=0; i<s.length(); i++) {
            charCounts[s.charAt(i) - 'a']++;
        }

        for (int i=0; i<s.length(); i++) {
            if (charCounts[s.charAt(i) - 'a'] == 1) {
                return i;
            }
        }

        return -1;
    }

    public static void main(String args[]){
        String s = "loveleetcode";
        System.out.println(firstUniqChar(s));
    }
}

// Time Complexity: O(n)
// Space Complexity: O(1)

File: 2-strings/cpp/387-first-unique-character.cpp
#include <bits/stdc++.h>
using namespace std;

int firstUniqChar(string s) {
    map <char, int> mp;

    for (int i=0; i<s.size(); i++) {
        mp[s[i]]++;
    }

    for (int i=0; i<s.size(); i++) {
        if (mp[s[i]] == 1) {
            return i;
        }
    }

    return -1;
}

int main () {
    string s = "loveleetcode";
    cout << firstUniqChar(s) << endl;

    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(n)